File: Tracing/Kernels.cs
﻿using Template;
using Cloo;
using System;

namespace clrays
{
    public interface VoidKernel<T> where T : struct
    {
        void Execute(ComputeEventList events);
        OpenCLBuffer<T> GetBuffer();
    }

    public interface ResultKernel<T> where T: struct
    {
        void Execute(ComputeEventList events);
        T[] GetResult();
        OpenCLBuffer<T> GetBuffer();
    }

    public class TraceAaKernel : ResultKernel<float>
    {
        private OpenCLKernel kernel;
        private readonly long[] work;
        private readonly float[] data;
        private readonly OpenCLBuffer<float> buffer;
        private readonly OpenCLBuffer<int> scene_params;
        private readonly OpenCLBuffer<float> scene_items;
        private bool dirty;

        public TraceAaKernel(OpenCLProgram program, Scene scene, uint width, uint height, uint AA)
        {
            data = new float[width * height * 3];
            buffer = new OpenCLBuffer<float>(program, data);
            //make kernel and set args
            kernel = new OpenCLKernel(program, "render");
            kernel.SetArgument(0, buffer);
            //make scene buffers
            var scene_raw = scene.GetBuffers();
            var scene_params_raw = scene.GetParamsBuffer();
            scene_params = new OpenCLBuffer<int>(program, scene_params_raw);
            scene_items = new OpenCLBuffer<float>(program, scene_raw);
            //set constants
            kernel.SetArgument(1, width);
            kernel.SetArgument(2, height);
            kernel.SetArgument(3, AA);
            //set arrays
            kernel.SetArgument(4, scene_params);
            kernel.SetArgument(5, scene_items);
            //work
            work = new long[] { width * AA, height * AA };
            dirty = false;
        }

        public void Execute(ComputeEventList events)
        {
            kernel.Execute(work, events);
            dirty = true;
        }

        public float[] GetResult()
        {
            if(dirty)
                buffer.CopyFromDevice();
            dirty = false;
            return data;
        }

        public OpenCLBuffer<float> GetBuffer()
        {
            return buffer;
        }
    }

    public class ClearKernel : VoidKernel<float>
    {
        private OpenCLKernel kernel;
        private readonly OpenCLBuffer<float> buffer;
        private readonly long[] work;

        public ClearKernel(OpenCLProgram program, OpenCLBuffer<float> buffer, uint width, uint height)
        {
            kernel = new OpenCLKernel(program, "clear");
            kernel.SetArgument(0, buffer);
            kernel.SetArgument(1, width);
            kernel.SetArgument(2, height);
            work = new long[] { width, height };
            this.buffer = buffer;
        }

        public void Execute(ComputeEventList events)
        {
            kernel.Execute(work, events);
        }

        public OpenCLBuffer<float> GetBuffer()
        {
            return buffer;
        }
    }

    public class ImageKernel : ResultKernel<int>
    {
        private OpenCLKernel kernel;
        private readonly long[] work;
        private readonly int[] data;
        private OpenCLBuffer<int> buffer;
        private bool dirty;

        public ImageKernel(OpenCLProgram program, OpenCLBuffer<float> input, uint width, uint height)
        {
            data = new int[width * height];
            buffer = new OpenCLBuffer<int>(program, data);
            kernel = new OpenCLKernel(program, "image_from_floatmap");
            kernel.SetArgument(0, input);
            kernel.SetArgument(1, buffer);
            kernel.SetArgument(2, width);
            kernel.SetArgument(3, height);
            work = new long[] { width, height };
            dirty = false;
        }

        public void Execute(ComputeEventList events)
        {
            kernel.Execute(work, events);
            dirty = true;
        }

        public int[] GetResult()
        {
            if (dirty)
                buffer.CopyFromDevice();
            dirty = false;
            return data;
        }

        public OpenCLBuffer<int> GetBuffer()
        {
            return buffer;
        }
    }
}

File: clrays-rs/src/kernels.rs
use ocl::{Kernel,Program,Queue};
use std::rc::Rc;
use crate::cl_helpers::{ClBuffer};
use crate::scene::Scene;
use crate::info::Info;

pub trait VoidKernel<T: ocl::OclPrm>{
    fn execute(&mut self, queue: &Queue) -> Result<(), ocl::Error>;
    fn get_buffer(&self) -> &ClBuffer<T>;
}

pub trait ResultKernel<T: ocl::OclPrm>{
    fn get_result(&mut self, queue: &Queue) -> Result<&[T],ocl::Error>;
}

pub struct ClearKernel{
    kernel: Kernel,
    buffer: Rc<ClBuffer<f32>>,
}

impl ClearKernel{
    pub fn new(name: &str, (w,h): (u32,u32), program: &Program, queue: &Queue, buffer: Rc<ClBuffer<f32>>) -> Result<Self,ocl::Error>{
        let kernel = unpack!(Kernel::builder()
        .program(program)
        .name(name)
        .queue(queue.clone())
        .global_work_size([w,h])
        .arg(buffer.get_ocl_buffer())
        .arg(&10f32)
        .build());
        Result::Ok(Self{ kernel,buffer })
    }
}

impl VoidKernel<f32> for ClearKernel{
    fn execute(&mut self, queue: &Queue) -> Result<(), ocl::Error>{
        unsafe { 
            match self.kernel.cmd().queue(queue).enq(){
                Ok(_) => Ok(()),
                Err(e) => Err(e),
            }
        }
    }

    fn get_buffer(&self) -> &ClBuffer<f32>{
        &self.buffer
    }
}

pub struct ImageKernel{
    kernel: Kernel,
    buffer: ClBuffer<i32>,
    dirty: bool,
}

impl ImageKernel{
    pub fn new(name: &str, (w,h): (usize,usize), program: &Program, queue: &Queue, input: &ClBuffer<f32>) -> Result<Self,ocl::Error>{
        let dirty = false;
        let buffer = match ClBuffer::<i32>::new(queue, w * h, 0){
            Ok(x) => x,
            Err(e) => return Err(e),
        };
        let kernel = unpack!(Kernel::builder()
        .program(program)
        .name(name)
        .queue(queue.clone())
        .global_work_size([w,h])
        .arg(input.get_ocl_buffer())
        .arg(buffer.get_ocl_buffer())
        .arg(w as u32)
        .arg(h as u32)
        .build());
        Ok(Self{ buffer,kernel,dirty })
    }
}

impl VoidKernel<i32> for ImageKernel{
    fn execute(&mut self, queue: &Queue) -> Result<(), ocl::Error>{
        unsafe { 
            match self.kernel.cmd().queue(queue).enq(){
                Ok(_) => { self.dirty = true; Ok(()) },
                Err(e) => Err(e),
            }
        }
    }

    fn get_buffer(&self) -> &ClBuffer<i32>{
        &self.buffer
    }
}

impl ResultKernel<i32> for ImageKernel{
    fn get_result(&mut self, queue: &Queue) -> Result<&[i32],ocl::Error>{
        if self.dirty {
            match self.buffer.download(queue){
                Ok(_) => {},
                Err(e) => return Err(e),
            }
        }
        self.dirty = false;
        Ok(self.buffer.get_slice())
    }
}

pub struct TraceKernel{
    kernel: Kernel,
    dirty: bool,
    buffer: ClBuffer<i32>,
    scene_params: ClBuffer<i32>,
    res: (u32,u32),
}

impl TraceKernel{
    pub fn new(name: &str, (w,h): (u32,u32), program: &Program, queue: &Queue, scene: &mut Scene, info: &mut Info) -> Result<Self, ocl::Error>{
        info.set_time_point("Start constructing TraceKernel");
        let dirty = false;
        let buffer = unpack!(ClBuffer::<i32>::new(queue, w as usize * h as usize, 0));
        info.set_time_point("Build int frame buffer");
        let scene_raw = scene.get_buffers();
        let scene_params_raw = scene.get_params_buffer();
        info.set_time_point("Build scene data");
        let mut scene_params = unpack!(ClBuffer::from(queue, scene_params_raw));
        let mut scene_items = unpack!(ClBuffer::from(queue, scene_raw));
        info.set_time_point("Build scene buffers");
        let tex_raw = scene.get_textures_buffer();
        let tex_params_raw = scene.get_texture_params_buffer();
        info.set_time_point("Build texture data");
        let mut tex_params = unpack!(ClBuffer::from(queue, tex_params_raw));
        let mut tex_items = unpack!(ClBuffer::from(queue, tex_raw));
        info.set_time_point("Build texture buffers");
        let kernel = unpack!(Kernel::builder()
        .program(program)
        .name(name)
        .queue(queue.clone())
        .global_work_size([w,h])
        .arg(buffer.get_ocl_buffer())
        .arg(w as u32)
        .arg(h as u32)
        .arg(scene_params.get_ocl_buffer())
        .arg(scene_items.get_ocl_buffer())
        .arg(tex_params.get_ocl_buffer())
        .arg(tex_items.get_ocl_buffer())
        .build());
        info.set_time_point("Create kernel");
        /*Choice: Either: upload an let ClBuffer's go out of scope
        Or: store ClBuffer's in the struct so they still live.
        They will be uploaded automatically then.
        I choose to upload here and let them go, as i don't need them later on and i can time the uploading.
        Except tehe scene_params. It is small and used to change camera etc. */
        unpack!(scene_params.upload(&queue));
        info.set_time_point("Upload scene_params");
        unpack!(scene_items.upload(&queue));
        info.set_time_point("Upload scene_items");
        unpack!(tex_params.upload(&queue));
        info.set_time_point("Upload tex_params");
        unpack!(tex_items.upload(&queue));
        info.set_time_point("Upload tex_items");
        Ok(Self{ kernel, dirty, buffer, scene_params, res: (w,h) })
    }

    pub fn update(&mut self, queue: &Queue) -> Result<(),ocl::Error>{
        unpack!(self.scene_params.upload(queue));
        Ok(())
    }

    pub fn get_res(&self) -> (u32,u32){
        self.res
    }
}

impl VoidKernel<i32> for TraceKernel{
    fn execute(&mut self, queue: &Queue) -> Result<(), ocl::Error>{
        unsafe { 
            match self.kernel.cmd().queue(queue).enq(){
                Ok(_) => { self.dirty = true; Ok(()) },
                Err(e) => Err(e),
            }
        }
    }

    fn get_buffer(&self) -> &ClBuffer<i32>{
        &self.buffer
    }
}

impl ResultKernel<i32> for TraceKernel{
    fn get_result(&mut self, queue: &Queue) -> Result<&[i32],ocl::Error>{
        if self.dirty {
            match self.buffer.download(queue){
                Ok(_) => {},
                Err(e) => return Err(e),
            }
        }
        self.dirty = false;
        Ok(self.buffer.get_slice())
    }
}
